lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -390,29 +1075,111 @@ SILValue DifferentiationTransformer::promoteToDifferentiableFunction(
+        *this, builder, desiredIndices, derivativeFnKind, origFnOperand,
+        invoker, newBuffersToDealloc);
+    // Show an error at the operator, highlight the argument, and show a note
+    // at the definition site of the argument.
+    if (!derivativeFnAndIndices)
+      return nullptr;
+
+    auto derivativeFn = derivativeFnAndIndices->first;
+    context.recordGeneratedFunctionReference(derivativeFn);
+
+    // If desired indices are a subset of actual indices, create a "subset
+    // indices thunk" and destroy the emitted derivative function reference.
+    // - For JVPs: the thunked JVP returns a differential taking fewer
+    //   parameters (using `.zero` for the dropped parameters).
+    // - For VJPs: the thunked VJP returns a pullback that drops the unused
+    //   tangent values.
+    auto actualIndices = derivativeFnAndIndices->second;
+    // NOTE: `desiredIndices` may come from a partially-applied function and
+    // have smaller capacity than `actualIndices`. We expect this logic to go
+    // away when we support `@differentiable` partial apply.
+    // if (actualIndices != desiredIndices) { // TODO: Re-enable.
+    auto extendedDesiredIndices = desiredIndices.parameters->extendingCapacity(
+        context.getASTContext(), actualIndices.parameters->getCapacity());
+    if (actualIndices.source != desiredIndices.source ||
+        !actualIndices.parameters->equals(extendedDesiredIndices)) {
+      // Destroy the already emitted derivative function reference because it
+      // is no longer used.
+      builder.emitDestroyValueOperation(loc, derivativeFn);
+      // Check if underlying original function reference has been partially
+      // applied with arguments. If so, produce an error: parameter subset
+      // thunks do not yet support this case because partially applied arguments
+      // cannot be propagated to parameter subset thunks.
+      auto didPartiallyApplyArguments = [](SILValue original) {
+        while (auto *pai =
+                   peerThroughFunctionConversions<PartialApplyInst>(original)) {
+          if (pai->getNumArguments() > 0)
+            return true;
+          original = pai->getCallee();
+        }
+        return false;
+      };
+      if (didPartiallyApplyArguments(origFnOperand)) {
+        context.emitNondifferentiabilityError(
+            origFnOperand, invoker,
+            diag::autodiff_cannot_param_subset_thunk_partially_applied_orig_fn);
+        return nullptr;
+      }
+      // Create the parameter subset thunk.
+      assert(actualIndices.parameters->isSupersetOf(extendedDesiredIndices));
+      SILFunction *thunk;
+      SubstitutionMap interfaceSubs;
+      SILOptFunctionBuilder fb(transform);
+      std::tie(thunk, interfaceSubs) =
+          getOrCreateSubsetParametersThunkForDerivativeFunction(
+              fb, origFnOperand, derivativeFn, derivativeFnKind, desiredIndices,
+              actualIndices);
```

Comment:
Note: subset parameters thunk tests are blocked by derivative function emitters, which will be upstreamed next.